### PR TITLE
IBX-2334: Added remoteId support in export and update processes

### PR DESCRIPTION
**Pull request by ciastektk**
| Question                                  | Answer
| ---------------------------------------- | ------------------
| **JIRA issue**                          | [IBX-2334](https://issues.ibexa.co/browse/IBX-2334)
| **Type**                                   | feature
| **Target Ibexa DXP version** | `v3.3.15`, `v4.0.2`
| **BC breaks**                          | no
| **Doc needed**                       | no

List of changes:

- Move REST controllers to REST directory to have more clean structure - this is not a BC because these actions are for internal usage 
- Added `ContentController::getContentByIdAction`
- Added `ContentController::getContentByRemoteIdAction`
- Added remoteId support for content update action (`ContentHelper::getContentUri()`)
- Added remoteId support for content export (`ContentService::setContent()`)
- Added dedicated `ContentQueryType` to get content based on `remoteId` or `id`
- Dropped useless code

#### Checklist:
- [x] Provided PR description.
- [x] Tested the solution manually.
- [ ] Provided automated test coverage.
- [x] Checked that target branch is set correctly (master for features, the oldest supported for bugs).
- [x] Ran PHP CS Fixer for new PHP code (use `$ composer fix-cs`).
- [x] Asked for a review (ping `@ezsystems/php-dev-team`).

### CODE REVIEW

**Review by Nattfarinn — APPROVED** [review]

File: src/bundle/Controller/REST/ContentController.php [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+use EzSystems\EzRecommendationClient\Value\Content;
+use EzSystems\EzRecommendationClient\Value\ContentData;
+use Ibexa\Personalization\QueryType\ContentQueryType;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+
+final class ContentController extends RestController
+{
+    /** @var \eZ\Publish\API\Repository\Repository */
+    protected $repository;
```
Comment (adamwojs) [inline src/bundle/Controller/REST/ContentController.php]: Why `protected`instead of `private` ?

**Review by konradoboza — APPROVED** [review]

File: src/bundle/Controller/REST/ContentController.php (hunk 1 above)
Comment (ciastektk) [reply src/bundle/Controller/REST/ContentController.php]: This is required by `EzSystems\EzPlatformRest\Server\Controller` If we change it to private then we got:  "Access level to 'ContentController::$repository' must be 'protected' (as in class 'Controller') or weaker"

File: src/lib/QueryType/ContentQueryType.php [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+            ->setAllowedTypes('language', ['null', 'string']);
+    }
+
+    /**
+     * @phpstan-param array{
+     *  'contentId': int|string,
+     *  'language': ?string,
+     * } $parameters
+     *
+     * @return \eZ\Publish\API\Repository\Values\Content\Query
```
Comment (alongosz) [inline src/lib/QueryType/ContentQueryType.php]: This is redundant.

File: src/lib/QueryType/ContentQueryType.php [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+                    'language' => null,
+                ]
+            )
+            ->setAllowedTypes('contentId', ['int', 'string'])
+            ->setAllowedTypes('language', ['null', 'string']);
+    }
+
+    /**
+     * @phpstan-param array{
+     *  'contentId': int|string,
```
Comment (alongosz) [inline src/lib/QueryType/ContentQueryType.php]: Content item ID is always `int`
```suggestion
     *  'contentId': int,
```

**Review by webhdx — APPROVED** [review]

File: src/lib/QueryType/ContentQueryType.php [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+        $language = $parameters['language'];
+
+        $query->filter = new Criterion\LogicalAnd($this->buildCriteria($contentId, $language));
+
+        return $query;
+    }
+
+    /**
+     * @param int|string $contentId
+     *
```
Comment (alongosz) [inline src/lib/QueryType/ContentQueryType.php]:
```suggestion
```

Use strict type instead ...

File: src/lib/QueryType/ContentQueryType.php [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+     * @return array<\eZ\Publish\API\Repository\Values\Content\Query\Criterion>
+     */
+    private function buildCriteria($contentId, ?string $language = null): array
```
Comment (alongosz) [inline src/lib/QueryType/ContentQueryType.php]: ... here:
```suggestion
    private function buildCriteria(int $contentId, ?string $language = null): array
```

File: src/lib/QueryType/ContentQueryType.php [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+    protected function configureOptions(OptionsResolver $optionsResolver): void
+    {
+        $optionsResolver
+            ->setDefaults(
+                [
+                    'contentId' => null,
+                    'language' => null,
+                ]
+            )
+            ->setAllowedTypes('contentId', ['int', 'string'])
```
Comment (alongosz) [inline src/lib/QueryType/ContentQueryType.php]: possibly also
```suggestion
            ->setAllowedTypes('contentId', ['int'])
```
but if this is passed by Query Type from core then keep it and cast properly when building criteria

File: src/lib/Service/ContentService.php [hunk 7]
```diff
@@ -166,7 +172,7 @@ public function setContent(CoreContent $content, ContentOptions $options): array
         ];
 
         return [
-            'contentId' => $content->id,
+            'contentId' => $this->repositoryConfigResolver->useRemoteId() ? $contentInfo->remoteId : $content->id,
```
Comment (alongosz) [inline src/lib/Service/ContentService.php:175]: Okay... So now I understand union type issues earlier. This is not the cleanest architecture. You should have separate fields for ID and RemoteID. Those are not the same by their meaning nor strict type.

**Review by alongosz — CHANGES_REQUESTED** [review]

File: src/bundle/Controller/REST/ContentController.php (hunk 1 above)
Comment (Steveb-p) [reply src/bundle/Controller/REST/ContentController.php]: If so, why do we even declare this property though? :thinking:

File: src/lib/Helper/ContentHelper.php [hunk 8]
```diff
@@ -18,47 +18,57 @@
...
+    /** @var \EzSystems\EzRecommendationClient\Helper\ContentTypeHelper */
+    private $contentTypeHelper;
+
     /** @var \eZ\Publish\API\Repository\LocationService */
     private $locationService;
 
-    /** @var \eZ\Publish\Core\MVC\ConfigResolverInterface */
-    private $configResolver;
+    /** @var \Psr\Log\LoggerInterface */
+    private $logger;
```
Comment (Steveb-p) [inline src/lib/Helper/ContentHelper.php]: Looks like `LoggerAwareInterface` / `LoggerAwareTrait` + phpdoc with property declaration (`@property $logger LoggerInterface`) should be used instead.

File: src/lib/Service/ContentService.php (hunk 7 above)
Comment (ciastektk) [reply src/lib/Service/ContentService.php:175]: This is required by reco engine. Content is identified by `contentId` field. Sorry, I forget to mention about this in PR description.

File: src/bundle/Controller/REST/ContentController.php (hunk 1 above)
Comment (ciastektk) [reply src/bundle/Controller/REST/ContentController.php]: There seems is no explanation why this property is here. I will remove it.

File: src/lib/Helper/ContentHelper.php (hunk 8 above)
Comment (ciastektk) [reply src/lib/Helper/ContentHelper.php]: Fixed: https://github.com/ezsystems/ezrecommendation-client/pull/106/commits/dfb3372534f0a0eda284f591fe97e43ec1141b82

File: src/bundle/Controller/REST/ContentController.php (hunk 1 above)
Comment (ciastektk) [reply src/bundle/Controller/REST/ContentController.php]: Fixed: https://github.com/ezsystems/ezrecommendation-client/pull/106/commits/104a30c37169ef64d42fc4fa1229e13a65432a81

**Review by alongosz — APPROVED** [review]

**Pull request merged**